feat: convert projects to Playwright

Sembauke opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

<!-- Feel free to add any additional description of changes below this line -->

Sembauke committed:
feat: convert projects to Playwright

Sembauke committed:
fix: disable bracket duplication when testing

Sembauke committed:
fix: login

Sembauke committed:
Merge branch 'main' into feat/convert-projects-to-playwright

Sembauke committed:
fix: add new editor helpers and fix test for fireFox

Sembauke committed:
feat: projects are submitted an viewed correctly test

Sembauke committed:
fix: only test directly one time

Sembauke committed:
feat: make test faster by not parsing the entire curriculum

Sembauke committed:
fix: actually claim cert

huyenltnguyen committed:
wip: fix failing tests on mobile device

Sembauke commented:
> The tests are still failing because there is a stray } in the editor, causing the code to be invalid and blocking the submission. The solution code in the JSON file looks correct to me, so I'm still (pulling my hair) trying to figure out where that character comes from (both editor.evaluate() and editor.fill() gave the same result).

You deleted the parameter which disables bracket and quote completion in the editor.

<img width="916" alt="image" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/46919888/0306b540-f58e-4879-81b0-b40b3d20fa2f">

This function has been added as even when filling or pasting code the editor completes brackets:

```
    // disable bracket completion when testing as it duplicates brackets when filling text
    const disableBracketCompletion =
      window.location.href.includes('testing=true');

    console.log(window.location.href, disableBracketCompletion);

    if (disableBracketCompletion) {
      editor?.updateOptions({
        autoClosingBrackets: 'never',
        autoClosingQuotes: 'never'
      });
    }
```

I should have left a comment about it.

huyenltnguyen committed:
revert the URL change

huyenltnguyen committed:
wait for solution modal to close

Sembauke committed:
fix: editor values for firefox and seed certified after all test

ojeytonwilliams commented:
@huyenltnguyen yep, it's the cookies. The issue is that the cookies for `certifieduser` are used even after the db is seeded with `developmentuser`. The fix/workaround is to create storageStates for both so that you can switch between users.

I'll create a quick PR for that, so you can use it here.

Sembauke committed:
feat: disable testing with Electron

ojeytonwilliams commented:
https://github.com/freeCodeCamp/freeCodeCamp/pull/54752 should help.

Once it's in, switching users means both changing storage state and seeding the desired user.

Sembauke committed:
Merge branch 'main' into feat/convert-projects-to-playwright

Sembauke committed:
fix: seeding possibly?

huyenltnguyen committed:
keep using dev user storageState after each test

ojeytonwilliams commented:
> I think the second test.use runs before the tests and overrides the first test.use, which is why the conflict happened

Pretty much. It's not exactly a conflict, it's just that multiple `test.use` calls override each other. As far as I understand it, `test.use` simply set the Playwright config for that scope (the entire file or a single describe block), so making two calls in the same scope is equivalent to just making the second call.

Sembauke committed:
feat: push for debug purposes

Sembauke committed:
fix: get solution to the bare minimum to pass the tests

Sembauke committed:
feat: final test

ojeytonwilliams commented on e2e/projects.spec.ts:
```diff
@@ -0,0 +1,276 @@
+  test.beforeAll(() => {
+    execSync('node ./tools/scripts/seed/seed-demo-user');
+  });
+
```
```suggestion
  test.skip(
    ({ browserName }) => browserName !== 'chromium',
    'Only chromium allows us to use the clipboard API.'
  );

```
So we can paste below.

ojeytonwilliams commented on e2e/projects.spec.ts:
```diff
@@ -0,0 +1,276 @@
+    // Use the `testing=true` parameter to prevent the editor
+    // from automatically adding a closing bracket.
+    await page.goto(
+      '/learn/javascript-algorithms-and-data-structures/javascript-algorithms-and-data-structures-projects/palindrome-checker?testing=true'
```
```suggestion
      '/learn/javascript-algorithms-and-data-structures/javascript-algorithms-and-data-structures-projects/palindrome-checker'
```
It won't be necessary if we use the clipboard.

ojeytonwilliams commented on e2e/projects.spec.ts:
```diff
@@ -0,0 +1,276 @@
+    await focusProjectEditor({ page, browserName, isMobile });
+    await clearProjectEditor({ page, browserName });
+
+    // To pass on firefox, we need to make sure the editor value is only set once
+    // instead of multiple times. (Bug in firefox)
+    await editor.evaluate((element, value) => {
+      (element as HTMLTextAreaElement).value = value;
+      element.dispatchEvent(new Event('input', { bubbles: true }));
+    }, contents);
```
```suggestion
    await page.evaluate(
      async contents => await navigator.clipboard.writeText(contents),
      contents
    );
    // Only runs on Chromium, so Control will work.
    await page.keyboard.press('Control+v');
```

Sembauke committed:
Apply suggestions from code review

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

huyenltnguyen committed:
grant clipboard permissions

huyenltnguyen committed:
use Meta+v if the test is run on MacOS

huyenltnguyen committed:
call authedPost concurrently

huyenltnguyen committed:
use getBy selectors

huyenltnguyen committed:
remove unused constant

huyenltnguyen committed:
use copy-paste to speed up the test

huyenltnguyen committed:
run the entire test suite as development user

huyenltnguyen committed:
ensure the show cert link has correct URL

huyenltnguyen committed:
remove similar test Cypress case

huyenltnguyen committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into pr/54689

ojeytonwilliams committed:
refactor: use one function for all requests

huyenltnguyen committed:
use getByLabel

huyenltnguyen committed:
rename editor.ts to challenge-editor.ts

huyenltnguyen committed:
remove unused prop

huyenltnguyen reviewed (approved):
Hookay, I think the PR is good to go.

This is one heck of a test suite to write / rewrite 😰

huyenltnguyen committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into pr/54689

huyenltnguyen committed:
update import statement in lower-jaw.spec.ts

ojeytonwilliams committed:
refactor: use one editor helper

It works for both projects and normal lessons, so there's no need to
handle them separately.

ojeytonwilliams commented on e2e/utils/request.ts:
```diff
@@ -3,24 +3,35 @@ import { APIRequestContext, expect } from '@playwright/test';
       headers: { 'csrf-token': csrfToken! }
     }
   );
+
+  if (res.status() !== 200) {
+    console.log(await res.json());
+  }
+
```
```suggestion
```
We can just look in the network tab if there's an error.

ojeytonwilliams committed:
refactor: remove console.log

ojeytonwilliams commented on client/src/templates/Challenges/classic/editor.tsx:
```diff
@@ -309,6 +309,7 @@ const Editor = (props: EditorProps): JSX.Element => {
       props.challengeType !== challengeTypes.multifilePythonCertProject
         ? 2
         : 4,
+
```
```suggestion
```

ojeytonwilliams committed:
refactor: remove extra spaces

ojeytonwilliams committed:
fix: editor helper

ojeytonwilliams reviewed (approved):
This was a struggle. Nice work, @Sembauke and @huyenltnguyen!

LGTM :+1:
